### ISSUE TITLE
Handle files

### ISSUE THREAD

**Issue by bickelj** [PhilanthropyDataCommons/service#216]
The PDC needs to accept file data, e.g. pdf, somehow.

Options discussed in the past include:
1. S3 or S3-like storage, with URLs in the database field values,
2. storage in the database itself,
3. or storage on a host filesystem.

Furthermore, in order for those files to be useful, they need to be readable by authenticated (and authorized) users.

**Comment by bickelj** [comment]
In the Zulip discussion at https://chat.opentechstrategies.com/#narrow/stream/66-PDC/topic/files, my current thought is to store URLs in the database field values, but that file attachments (both read and write) be handled by this PDC service (rather than attempting to offload it to S3 in a matter exposed to clients). It might be OK for the service on the back-end to use S3 or a filesystem or whatever else, but let the files be an endpoint within the PDC service.

**Comment by bickelj** [comment]
Something like one of these flows:

1. POST a proposal, response returns id.
2. POST attachment(s) using the proposal id from (1), response returns URL(s).
3. POST proposal version with proposal id from (1) and field values containing such URLs from (2).

Or

1. POST a proposal, response returns id.
2. POST proposal version with all non-file fields using the proposal id from (1).
3. POST attachment(s) using the proposal id from (1), which automatically generates new proposal versions having new field values containing such URLs that are created via the POSTs here in (3).

**Comment by jasonaowen** [comment]
I strongly suggest not attempting to use the database as a file storage server, nor to store any data on the instance - S3 or something like it is definitely the way to go. Note that [Digital Ocean has S3-compatible storage](https://www.digitalocean.com/products/spaces).

Uploading a file would involve asking the API for a presigned URL, uploading to that URL, and then informing the API that the upload completed successfully. Uploads would go to a prefix in the bucket that had an automatic expiration policy to delete partial, failed, or incomplete uploads; during processing, the API would move the attachment from the automatically-generated placeholder filename (such as a UUID) to its permanent name, outside of the expiration policy prefix.

This is a standard upload flow in most every web application I've encountered, and @slifty and I both have a fair amount of experience implementing it.

**Comment by bickelj** [comment]
See also https://github.com/PhilanthropyDataCommons/service/issues/396.

This issue is related in the sense that now there are (at least) two uses for uploading documents:
1. Attach a document to a proposal (above)
2. Upload raw proposal data in CSV form, process it into JSON, further process it into the PDC database (https://github.com/PhilanthropyDataCommons/service/issues/102, other stuff from Phase II)

**Comment by bickelj** [comment]
There is an authorization question that will drive the design.

If I copy a link to a document within PDC and send it to another person, should that person be required to authenticate to see the document or is it OK for the link to contain a secret that exposes the document to whomever has the secret link?

I personally would not upload documents in the latter case. I think the document needs to be protected. In other words, no direct S3 access unless the S3 system asks our auth service if the current authentication session is valid.

**Comment by bickelj** [comment]
I doubt there is a satisfactory S3 integration with OpenID Connect. Both Ceph and minio, for example, use a "validate the JWT using the auth service and then issue temporary storage-system-specific credentials" model. I suppose it "works" but seems clunky to me.

From https://min.io/docs/minio/linux/administration/identity-access-management/oidc-access-management.html#id1
> If the JWT is valid, MinIO checks for a [claim](https://min.io/docs/minio/linux/administration/identity-access-management/oidc-access-management.html#minio-external-identity-management-openid-access-control) specifying a list of one or more [policies](https://min.io/docs/minio/linux/administration/identity-access-management/policy-based-access-control.html#minio-policy) to assign to the authenticated user. MinIO defaults to checking the policy claim. MinIO returns temporary credentials in the STS API response in...

From https://docs.ceph.com/en/latest/radosgw/keycloak/#integrating-keycloak-with-radosgw
> By using the web token returned by the authentication process, a mobile app or web app can call AssumeRoleWithWebIdentity, receive a set of temporary S3 credentials, and use those credentials to make S3 calls.

**Comment by reefdog** [comment]
> If I copy a link to a document within PDC and send it to another person, should that person be required to authenticate to see the document or is it OK for the link to contain a secret that exposes the document to whomever has the secret link?

The former. The canonical URL for a resource protected by authorization should, by default, require authorization.

We can choose to support the latter, but it should require an explicit action by the sharing user (beyond copying the canonical URL from the address bar or file listing), such as clicking a button that reads "Generate shareable link". There should be explanatory text that explains anyone with the link can view the document. IMO this feature can wait to be built until we know we need it.

**Comment by slifty** [comment]
I *believe* we can have an endpoint that is protected by OpenID, which generates a presigned post that requires a particular header to be populated (that header could be something that is very secure) -- this seems like it would be good enough while also fitting nicely and elegantly in our existing auth flow.

**Comment by bickelj** [comment]
@slifty As you can tell, I haven't written to S3 in forever (and even when I did, I probably wasn't doing it correctly). I assume docs such as https://min.io/docs/minio/linux/integrations/presigned-put-upload-via-browser.html are relevant for my reading. I had interpreted this workflow as only applying to `POST` and `PUT`, but it looks like it also applies to `GET` as well. If the flow you described happens on each and every document request, has an S3 expiration matching that of the JWT, and requires the exact contents of the JWT to be correct, then my concern is addressed. Otherwise, yes, if you'd be willing to go into more detail that would help.

**Comment by reefdog** [comment]
@bickelj Semantically, I consider the `api.*` version the canonical URL since that's the stable URL we'd ideally support in perpetuity, while the S3¹ version is a current implementation detail.

But I was speaking in general design principles and my statement would equally apply to both versions: I don't think users should be able to **accidentally** share a pre-signed URL. I.e., any URL we expose to the user (through a file listing, through the address bar), if shared with an anonymous user, should not "just work" without further auth. If the user wants to share a file that "just works", they should have to take explicit action to generate the URL; having done so, I have no issue with a pre-signed URLs.

I've intentionally avoided technical implementation suggestions because I think you guys are in a better position to mull on those, but that's the UX that I would advocate for. If we can't escape pre-signed S3 URLs (with no header/other auth) as the final destination, then so be it; I would just hope we take pains to minimize the likelihood the user will see/share those URLs.

¹ Goes without saying that "S3" here means "whatever S3-protocol service we use".

**Comment by bickelj** [comment]
@reefdog Thanks for the clarification. I think you've well-stated the design principles and agree with almost all of it. The only reason I say "almost" is I would take it one step further. I think we ought to eliminate the likelihood the user will see/share those URLs, i.e. not even use those URLs in the user's agent with XHR such that a user could possibly see them.

**Comment by reefdog** [comment]
@bickelj I think that's a good goal though sometimes the medicine is worse than the disease. (This can be where things like full-viewport iframing come from, that make life worse for everyone but don't actually stop the motivated bad actor.) So I say let's move forward with making it unlikely the user accidentally does the wrong thing, and explore mechanisms for entirely preventing it.

**Comment by slifty** [comment]
> If the flow you described happens on each and every document request, has an S3 expiration matching that of the JWT, and requires the exact contents of the JWT to be correct, then my concern is addressed.

That is what I was thinking for sure and I do believe what you described is how we can get it to work!

**Comment by jasonaowen** [comment]
This is an apt question, and accurately describes one of the tradeoffs of using S3 (with whatever provider). To recap, sharing S3 URLs with the client has the drawback that for some (configurable, short) period of time, the presigned S3 URL we generate and return to the client could be accessed by anybody.

I think there are two options here:

1. Write a proxy that would serve files by downloading them from S3 and uploading them to the client
2. Accept the tradeoff and configure the URL signing to make the generated S3 URL only valid for 60 seconds (or whatever period we decide on)

I would argue for option 2; building out option 1 is a lot of engineering effort, and there are going to be pitfalls. If option 1 turns out later to be required, we can revisit this, but let's start with the quick, reasonably-secure approach!

**Comment by jasonaowen** [comment]
Also note that we cannot eliminate the risk of an authorized user sharing a file with someone who should not be able to access it: they can just download the file themselves and then share the file out-of-band. We can and should take steps to make sure that only authorized users are able to download files from our system, and limit the window of semi-inadvertent sharing, but without a compelling, specific security concern, I don't think this is worth spending too much time on.

**Comment by jasonaowen** [comment]
Here's a sequence diagram of how uploads would work, courtesy of [this StackOverflow answer about client-to-S3 uploads](https://stackoverflow.com/a/36202191):

![image](https://github.com/PhilanthropyDataCommons/service/assets/1494855/fdd79bc8-6be5-4486-9177-2b775f58092e)

(note that, as previously mentioned, we would not be using AWS to provide S3, but rather another S3 provider - probably [DigitalOcean Spaces](https://www.digitalocean.com/products/spaces), although we haven't discussed that yet.)

The download case is similar; if it would be helpful I can make a specific sequence diagram for downloads, but I wasn't able to find an existing one with a quick search.

Crucially, steps 1 and 2 apply to both uploads and downloads - whatever authorization we build out for #255 would ensure that we only generate S3 signatures for authorized users.

**Comment by slifty** [comment]
Just to document it: we did decide on call today that we do want to use the S3 protocol for the file storage.  This doesn't weigh on the above discussion about access since I think that was all rooted in assumption of S3

**Comment by jasonaowen** [comment]
If we were to share signed S3 get URLs with the user, we would need to set the expiration date on the URL. Tuning that has some tradeoffs between security and functionality.

- Shorter times are more secure: they limit the amount of time the URL is valid for, and so narrow the window when someone other than the authorized user could access the file at that URL
- Longer times are more tolerant to clock skew between the server generating the signature and the S3 server validating the signature
- Longer times are more tolerant to slow clients

Assuming no clock skew (which is not a safe assumption), in our use case, I believe we should be able to get away with a seconds-long expiration date, as the time between generation and use should be minimal: user requests a file, server generates a URL, server redirects browser to the signed URL, browser immediately starts downloading it. Taking clock skew and slow browsers into account, an expiration time of low-single-digit minutes seems appropriate. I would suggest we start with 1 minute and go from there, with the intention of making this a tunable parameter that we can adjust if we get reports of problems downloading.

Note also that [the expiration date is checked at the beginning of the request](https://docs.aws.amazon.com/en_us/AmazonCloudFront/latest/DeveloperGuide/private-content-signed-urls.html#private-content-check-expiration) - the request does NOT need to complete before the expiration date. This may affect clients that attempt ranged/multi-part downloads of large files, and again is something that we could tune if and as we receive support reports.

**Comment by slifty** [comment]
> The existence of an expiration to a pre-signed URL rather than an unexpiring pre-signed URL is due to the URL providing anyone anywhere repeated access to the document. It is an admission that there is a security issue and the time limit is a mitigation. Ideally the time limit is 0.

Just wanted to note that tokens have an expiration date as well!  I think the signed URL expiration concept is more in line with that.

It is more secure to have keys that degrade after their useful life.  That doesn't mean that the key itself is so insecure that it doesn't warrant existing to provide practical utility.  I guess put another way: the most secure possible thing would be to provide absolutely no access whatsoever, but of course that would not be a useful tradeoff to make.

I just wanted to note that the "what do the big players do" is not quite what I meant with that exploration: I more meant that this is not a new concept and that thousands of organizations have thought about this problem and we would be wise to take some time to understand their verdicts.

I have not found any resources *yet* arguing that S3's implemented model for secure URLs is insecure.  That does not invalidate the perpsective of course, but it does signal that blunt consideration that they are an insecure concept (as opposed to a concept that is part of a nuanced landscape of threat models) is a fringe perspective even among security experts!  I'll keep looking, but I think it'd be helpful to ground our discussion in the work of others.

**Comment by slifty** [comment]
I'm working on an endpoint for file upload as part of the CDV bulk processing workflow -- #557

I suspect this endpoint will be relevant for this issue as well.

**Comment by bickelj** [comment]
From my (happily) fringe perspective, I think this issue has been at least partially resolved or superseded by #557/#567, #558/#628, #559/#658, etc.
